### ISSUE TITLE
nsenter: cannot open /proc/320/ns/time: No such file or directory

### ISSUE THREAD

**Issue by jototland** [DamionGans/ubuntu-wsl2-systemd-script#36]
I get this after running your script. 
nsenter: cannot open /proc/320/ns/time: No such file or directory

Luckily, I can "wsl -e bash --norc" and revert the changes to /etc/bash.bashrc. But what is wrong with my /proc since nsenter doesn't like it?

Windows version 19041.388
WSL 2
Pengwin

**Comment by jototland** [comment]
Ok, I tested again with ubuntu instead of pengwin, and the script worked just fine. I guess that might be the reason the script has ubuntu in it's name :-)

I would still like to make it work for pengwin, though...

**Comment by khs1994** [comment]
https://github.com/DamionGans/ubuntu-wsl2-systemd-script/blob/4d221c4f9bf89187eee7cd7ee298e5344190b191/enter-systemd-namespace#L41


this error because `nsenter` CMD -a maybe include `time namespaces`, you can  diff `ubuntu` and `pengwin` `nsenter --help`

you can change `-a` to `--mount --uts ...` 

> The time namespace support has been merged in the 5.6 kernel.

**Comment by eternalphane** [comment]
Please see https://github.com/yuk7/ArchWSL2/issues/21#issuecomment-663981023

**Comment by pablorq** [comment]
**TL;DR: Change options of nsenter from `-a` to `-m -p`**

Thank you @eternalphane for the information, but this could be confusing for some people.

Based on that data, this is how to fix it:

- Replace (copy and paste) the following lines of `enter-systemd-namespace` file:

``` bash
USER_HOME="$(getent passwd | awk -F: '$1=="'"$SUDO_USER"'" {print $6}')"
if [ -n "$SYSTEMD_PID" ] && [ "$SYSTEMD_PID" != "1" ]; then
    if [ -n "$1" ] && [ "$1" != "bash --login" ] && [ "$1" != "/bin/bash --login" ]; then
        exec /usr/bin/nsenter -t "$SYSTEMD_PID" -m -p \
            /usr/bin/sudo -H -u "$SUDO_USER" \
            /bin/bash -c 'set -a; [ -f "$HOME/.systemd-env" ] && source "$HOME/.systemd-env"; set +a; exec bash -c '"$(printf "%q" "$@")"
    else
        exec /usr/bin/nsenter -t "$SYSTEMD_PID" -m -p \
            /bin/login -p -f "$SUDO_USER" \
            $([ -f "$USER_HOME/.systemd-env" ] && /bin/cat "$USER_HOME/.systemd-env" | xargs printf ' %q')
    fi
    echo "Existential crisis"
    exit 1
fi
```

- Reintall it:
``` bash
bash ubuntu-wsl2-systemd-script.sh --force
```

- Worked on:
    - WSL2 + Ubuntu-20.10

**Comment by s4pfyr** [comment]
Thanks @pablorq !
This was needed for me to get in working on WSL2+Ubuntu after release upgrade to 21.04!

**Comment by pablorq** [comment]
I'm glad it helps!

:slightly_smiling_face:

**Comment by hyoretsu** [comment]
@s4pfyr How exactly did you do it? After upgrading, mine started throwing `Failed to enable unit: Transport endpoint is not connected` (Docker) and `nsenter: cannot open /proc/9286/ns/time: No such file or directory` when running `exec sudo nsenter -m -u -i -n -p -C -r -w -t $(pidof -s systemd) -a su - $LOGNAME`

**Comment by s4pfyr** [comment]
@hyoretsu I did not have the nsenter problem after switching the flags like @pablorq suggested.
Regarding docker: I had problems with iptables and needed to do `sudo update-alternatives --config iptables` and select legancy mode. Try starting the docker daemon with `sudo dockerd --debug` to figure out whats going on with docker.

**Comment by MaxG-Enpal** [comment]
I'm even unable to run the install script with --force:

nsenter: cannot open /proc/830/ns/time: No such file or directory

**Comment by hyoretsu** [comment]
> I'm even unable to run the install script with --force:
> 
> nsenter: cannot open /proc/830/ns/time: No such file or directory

There are two commands, they each work in different cases:
`exec sudo nsenter -m -u -i -n -p -C -r -w -t $(pidof systemd) -a su - $LOGNAME`
`exec sudo nsenter -m -u -i -n -p -C -r -w -t $(pidof systemd) -m -p su - $LOGNAME` (just noticed mp being repeated, might not be necessary)

If that doesn't work, I believe you need to do it [manually](https://github.com/microsoft/WSL/issues/5126#issuecomment-716389723) 

I managed to successfully update to 21.10 btw
https://github.com/microsoft/WSL/issues/5126#issuecomment-1001240706

**Comment by MaxG-Enpal** [comment]
> 

ok so have I understood it correctly.
I need to replace the LOGNAME with my home user name and pidof systemd.
But how can I find the PID of systemd if systemd doesnt work? (Or even if it works)

**Comment by hyoretsu** [comment]
> > 
> 
> ok so have I understood it correctly. I need to replace the LOGNAME with my home user name and pidof systemd. But how can I find the PID of systemd if systemd doesnt work? (Or even if it works)

`top systemd`/`htop systemd` though keep in mind that the PID changes sometimes.

**Comment by IgorOhrimenko** [comment]
I try 
```
wsl bash --norc
nsenter: cannot open /proc/26/ns/time: No such file or directory
```
it does not work, but
```
wsl --exec /usr/bin/sh
$
```
work and then I used @pablorq's method.

**Comment by IgorOhrimenko** [comment]
Check your the endfile /usr/sbin/enter-systemd-namespace, if need change it.

**Comment by eitanmiz** [comment]
As I seen in: https://www.csa.iisc.ac.in/~vg/teaching/E0-256/slides/lxc-slide-deck.pdf
 /proc/pid/ns/time is new, so I think that there may be patch for update, or just downgrade some services.

**Comment by DataMadeEEZY** [comment]
I ran this script ( ubuntu-wsl2-systemd-script) originally but did not have success. So I waited for upgrade to 22.04 like most suggested and after upgrading I run into this error:
 nsenter: cannot open /proc/27/ns/time: No such file or directory

[process exited with code 1 (0x00000001)]

 wsl -e /usr/libexec/nslogin /bin/bash in command prompt allows me to the enter the terminal. How can I get my linux terminal to open without error now?

**Comment by Pflegusch** [comment]
Thanks @pablorq, worked for me!

**Comment by apos** [comment]
Unfortunately only works partly. But the system starts, only it is not possible to login as normal user. 

`wsl -u root` works totally fine. 

`tail -f /var/log/syslog` gives 

    systemd-fstab-generator[14]: Failed to create unit file /run/systemd/generator/-.mount, as it already exists. Duplicate entry in /etc/fstab?
    systemd-sysv-generator[22]: Failed to create unit file /run/systemd/generator.late/hddtemp.service: File exists
    systemd-sysv-generator[22]: Failed to create unit file /run/systemd/generator.late/apport.service: File exists
    systemd-journald[32]: File /var/log/journal/cad812440ce02ca8f59723955f259f43/system.journal corrupted or uncleanly shut down, renaming and replacing.
    systemd-journald[32]: Failed to read journal file /var/log/journal/cad812440ce02ca8f59723955f259f43/user-1000.journal for rotation, trying to move it out of the way: Device or resource busy
    systemd-journald[32]: Failed to read journal file /var/log/journal/cad812440ce02ca8f59723955f259f43/user-1000.journal for rotation, trying to move it out of the way: Device or resource busy

**Comment by decli** [comment]
> ```shell
> USER_HOME="$(getent passwd | awk -F: '$1=="'"$SUDO_USER"'" {print $6}')"
> if [ -n "$SYSTEMD_PID" ] && [ "$SYSTEMD_PID" != "1" ]; then
>     if [ -n "$1" ] && [ "$1" != "bash --login" ] && [ "$1" != "/bin/bash --login" ]; then
>         exec /usr/bin/nsenter -t "$SYSTEMD_PID" -m -p \
>             /usr/bin/sudo -H -u "$SUDO_USER" \
>             /bin/bash -c 'set -a; [ -f "$HOME/.systemd-env" ] && source "$HOME/.systemd-env"; set +a; exec bash -c '"$(printf "%q" "$@")"
>     else
>         exec /usr/bin/nsenter -t "$SYSTEMD_PID" -m -p \
>             /bin/login -p -f "$SUDO_USER" \
>             $([ -f "$USER_HOME/.systemd-env" ] && /bin/cat "$USER_HOME/.systemd-env" | xargs printf ' %q')
>     fi
>     echo "Existential crisis"
>     exit 1
> fi
> ```

it works on wsl2 + Ubuntu 22.04.1

**Comment by danztensai** [comment]
https://github.com/DamionGans/ubuntu-wsl2-systemd-script/issues/36#issuecomment-1214179638 working for me !!! THANK"S you save my life

**Comment by Lestat06** [comment]
> bash ubuntu-wsl2-systemd-script.sh --force

@pablorq thanks for the save, however worked for me after running `bash ubuntu-wsl2-systemd-script.sh --force` with sudo

**Comment by cferrante** [comment]
Thanks @jototland for opening the ticket, and thanks @pablorq for the simplification of the fix!
Had to use this fix on WSL2+Ubuntu after release upgrade to 22.04.

**Comment by ervalade** [comment]
Hi guys
I've upgraded 20.04 to 22.04, 
I've got the problem with /proc/xx/ns/time,
I've changed the  enter-systemd-namespace ; the problem /proc ... was fixed
but when I start wsl,  
-----
Welcome to Ubuntu 22.04.1 LTS (GNU/Linux 5.15.79.1-microsoft-standard-WSL2 x86_64)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

  System information as of sam. 31 déc. 2022 18:33:48 CET

  System load:  0.830078125        Processes:             9
  Usage of /:   6.7% of 250.92GB   Users logged in:       0
  Memory usage: 4%                 IPv4 address for eth0: 172.23.118.163
  Swap usage:   0%

 * Strictly confined Kubernetes makes edge and IoT secure. Learn how MicroK8s
   just raised the bar for easy, resilient and secure K8s cluster deployment.

   https://ubuntu.com/engage/secure-kubernetes-at-the-edge

3 mises à jour peuvent être appliquées immédiatement.
Pour afficher ces mises à jour supplémentaires, exécuter : apt list --upgradable


Last login: Sat Dec 31 18:23:02 CET 2022 on pts/0
---------
then I've got "**Login timed out after 60 seconds.**"

after some investigations, i've found the command "getent password" never end
any idea ?
thanks

**Comment by ringerc** [comment]
Was fixed in nsenter, see https://github.com/util-linux/util-linux/issues/2059

**Comment by icrc-aasselineau** [comment]
Worked on WSL2 + Ubuntu 22.04, also
Thanks @pablorq

**Comment by 0jdk** [comment]
> **TL;DR: Change options of nsenter from `-a` to `-m -p`**
> 
> Thank you @eternalphane for the information, but this could be confusing for some people.
> 
> Based on that data, this is how to fix it:
> 
> * Replace (copy and paste) the following lines of `enter-systemd-namespace` file. You can do it on Windows side.
> 
> ```shell
> USER_HOME="$(getent passwd | awk -F: '$1=="'"$SUDO_USER"'" {print $6}')"
> if [ -n "$SYSTEMD_PID" ] && [ "$SYSTEMD_PID" != "1" ]; then
>     if [ -n "$1" ] && [ "$1" != "bash --login" ] && [ "$1" != "/bin/bash --login" ]; then
>         exec /usr/bin/nsenter -t "$SYSTEMD_PID" -m -p \
>             /usr/bin/sudo -H -u "$SUDO_USER" \
>             /bin/bash -c 'set -a; [ -f "$HOME/.systemd-env" ] && source "$HOME/.systemd-env"; set +a; exec bash -c '"$(printf "%q" "$@")"
>     else
>         exec /usr/bin/nsenter -t "$SYSTEMD_PID" -m -p \
>             /bin/login -p -f "$SUDO_USER" \
>             $([ -f "$USER_HOME/.systemd-env" ] && /bin/cat "$USER_HOME/.systemd-env" | xargs printf ' %q')
>     fi
>     echo "Existential crisis"
>     exit 1
> fi
> ```
> 
> * Get inside the broken Linux distribution:
> 
> ```
> > **wsl -e bash --norc**
> ```
> 
> * Reinstall it:
> 
> ```
> $ **sudo** bash ubuntu-wsl2-systemd-script.sh --force
> ```
> 
> * Worked on:
>   
>   * WSL2 + Ubuntu-20.10
